File: itx-rs/itx-contract/src/repo/post.rs
use crate::repo::error::RepoError;
use async_trait::async_trait;
use uuid::Uuid;

pub type PostId = i64;
pub type AuthorId = Uuid;

#[derive(Debug, Clone)]
pub struct Post {
    pub id: PostId,
    pub author_id: AuthorId,
    pub title: String,
    pub body: String,
    pub tags: Vec<String>,
    pub created_at: time::OffsetDateTime,
}

#[derive(Debug, Clone)]
pub struct NewPost {
    pub author_id: AuthorId,
    pub title: String,
    pub body: String,
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Default)]
pub struct PostPatch {
    pub title: Option<String>,
    pub body: Option<String>,
    pub tags: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default)]
pub struct ListPostsQuery {
    pub author_id: Option<AuthorId>,
    pub limit: u32,
    pub offset: u32,
}

#[async_trait]
pub trait PostRepo: Send + Sync {
    async fn list(&self, query: ListPostsQuery) -> Result<Vec<Post>, RepoError>;

    async fn get(&self, id: PostId) -> Result<Option<Post>, RepoError>;

    async fn create(&self, input: NewPost) -> Result<Post, RepoError>;

    /// Updates a post owned by `author_id`. Returns `None` if the post does
    /// not exist or is not owned by the caller — handlers should map both to 404.
    async fn update(
        &self,
        id: PostId,
        author_id: AuthorId,
        patch: PostPatch,
    ) -> Result<Option<Post>, RepoError>;

    /// Deletes a post owned by `author_id`. Returns `false` if the post does
    /// not exist or is not owned by the caller.
    async fn delete(&self, id: PostId, author_id: AuthorId) -> Result<bool, RepoError>;
}

File: itx-kt/itx-contract/src/main/kotlin/io/github/chehsunliu/itx/contract/repo/Post.kt
package io.github.chehsunliu.itx.contract.repo

import java.time.OffsetDateTime
import java.util.UUID

data class Post(
    val id: Long,
    val authorId: UUID,
    val title: String,
    val body: String,
    val tags: List<String>,
    val createdAt: OffsetDateTime,
)

interface PostRepo {
    suspend fun list(params: ListParams): List<Post>

    /** Throws [RepoNotFoundException] if no post with `params.id` exists. */
    suspend fun get(params: GetParams): Post

    suspend fun create(params: CreateParams): Post

    /**
     * Updates a post owned by `params.authorId`. Throws [RepoNotFoundException] if the post does
     * not exist or is not owned by the caller.
     */
    suspend fun update(params: UpdateParams): Post

    /**
     * Deletes a post owned by `params.authorId`. Throws [RepoNotFoundException] if the post does
     * not exist or is not owned by the caller.
     */
    suspend fun delete(params: DeleteParams)

    data class ListParams(
        val authorId: UUID? = null,
        val limit: Int = 0,
        val offset: Int = 0,
    )

    data class GetParams(
        val id: Long,
    )

    data class CreateParams(
        val authorId: UUID,
        val title: String,
        val body: String,
        val tags: List<String>,
    )

    data class UpdateParams(
        val id: Long,
        val authorId: UUID,
        val title: String? = null,
        val body: String? = null,
        val tags: List<String>? = null,
    )

    data class DeleteParams(
        val id: Long,
        val authorId: UUID,
    )
}
